Add Performance Tuning Guide

maximiliankolb opened the pull request:
This PR migrates the "Satellite Performance Tuning Guide" from https://github.com/RedHatSatellite/satellite-performance-tuning.

Discussion: https://github.com/RedHatSatellite/satellite-performance-tuning/issues/33


Cherry-pick into:

- [x] Foreman 3.3
- [x] Foreman 3.2
- [x] Foreman 3.1

maximiliankolb commented:
Current status: **draft**; I need to check the migrated content, esp. links, branding, and tables.

ekohl commented:
> Current status: **draft**; I need to check the migrated content, esp. links, branding, and tables.

When creating a PR you can mark it as draft. The UI has a button, [hub](https://github.com/github/hub) has a `-d` argument. I've now used the Convert to draft button. Feel free to hit `Ready for review` whenever you think it's ready. You can always convert it back to draft.

maximiliankolb commented:
**DO NOT MERGE**

This PR is ready for a first review. It's not completely done yet, but I need some input and feedback.

maximiliankolb commented:
Rebased to master. I don't know why it's not building. Any input would be greatly appreciated!

ekohl commented on guides/common/modules/proc_configuring-the-open-files-limit-for-apache-httpd.adoc:
```diff
@@ -0,0 +1,21 @@
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# systemctl daemon-reload
+# {foreman-maintain} service restart
```
I hate this if I'm honest. I really don't know why we don't just use
```suggestion
# systemtl restart httpd
```

There is 0 reason to restart all services and only takes a long time.

I think there is also an option like `--only` to foreman-maintain but really, we're already deep into systemd by editing an override file (it doesn't mention you need to create `/etc/systemd/system/httpd.service.d` for example). It also calls daemon-reload.

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/proc_configuring-the-qpid-mgmt-pub-interval-parameter.adoc:
```diff
@@ -0,0 +1,32 @@
+The old management object is not yet purged, which is why qpid throws this error.
+
+.Procedure
+. Set the `mgmt-pub-interval` parameter in `/etc/qpid/qpidd.conf`:
++
+[options="nowrap" subs="+quotes,attributes"]
+----
+mgmt-pub-interval=5
```
This is bad advise since the installer will revert this anyway. Instead, it should advise custom-hiera.yaml:
```yaml
qpid::mgmt_pub_interval: 5
```

Then rerun the installer.

maximiliankolb replied:
Thanks, applied!

ekohl commented on guides/common/modules/proc_enabling-tuned-profiles.adoc:
```diff
@@ -0,0 +1,42 @@
+
+{RHEL} 7 enables the tuned daemon by default during installation.
+On bare-metal, {Team} recommends to run the `throughput-performance` tuned profile on {ProjectServer} and {SmartProxies}.
+On virtual maschines, {Team} recommdsn to run the `virtual-guest` profile.
```
```suggestion
On virtual machines, {Team} recommends to run the `virtual-guest` profile.
```

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/proc_enabling-tuned-profiles.adoc:
```diff
@@ -0,0 +1,42 @@
+----
+# systemctl status tuned
+----
+. If `tuned` is not running, enable it:
++
+[options="nowrap" subs="+quotes,attributes"]
+----
+# systemctl enable tuned
```
`enable` does not start the service. It just enables it (which means: start it at boot). If you want to start it
```suggestion
. If `tuned` is not running, enable it:
+
[options="nowrap" subs="+quotes,attributes"]
----
# systemctl enable --now tuned
```

maximiliankolb replied:
Applied.

ekohl commented on guides/common/modules/proc_enabling-tuned-profiles.adoc:
```diff
@@ -0,0 +1,42 @@
+// what does this mean? Where should it go?
+Transparent Huge Pages is a memory management technique used by the Linux kernel which reduces the overhead of using Translation Lookaside Buffer (TLB) by using larger sized memory pages.
+Due to databases having Sparse Memory Access patterns instead of Contiguous Memory access patterns, database workloads often perform poorly when Transparent Huge Pages is enabled.
+To improve performance of PostgreSQL, {Team} recommends Transparent Huge Pages be disabled.
```
This isn't just us, it's really a best practice. https://www.percona.com/blog/2019/03/06/settling-the-myth-of-transparent-hugepages-for-databases/ actually has some nice write up.

maximiliankolb replied:
Reworded.

ekohl commented on guides/common/modules/con_puma-workers-and-threads-autotuning.adoc:
```diff
@@ -0,0 +1,35 @@
+Environment=FOREMAN_PUMA_THREADS_MIN=5
+Environment=FOREMAN_PUMA_THREADS_MAX=5
+Environment=FOREMAN_PUMA_WORKERS=30
+# pgrep -u foreman --list-full | grep 'puma: cluster worker'
```
Wouldn't it be easier to see the progress tree using `systemctl status foreman.service`?

maximiliankolb replied:
Yes, added.

ekohl reviewed (commented):
Not a complete review since I didn't read through the Puma since the recommendations changed.

maximiliankolb commented on guides/common/modules/con_performance-tuning-quickstart.adoc:
```diff
@@ -0,0 +1,33 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+# {foreman-installer} --tuning "_MY_INSTALLATION_SIZE_"
```
I've made this visible for ProjectServer only for Katello and downstream.

maximiliankolb commented on guides/common/modules/con_postgresql-tuning.adoc:
```diff
@@ -0,0 +1,42 @@
+
+The below set of tunings can be applied to PostgreSQL to improve its response times (see `how to use custom-hiera.yaml` file; this will modify `/var/lib/pgsql/data/postgresql.conf` file):
+
+----
```
Yes, added.

maximiliankolb commented on guides/common/modules/con_postgresql-tuning.adoc:
```diff
@@ -0,0 +1,42 @@
+The database sees an extensive usage is usually working on to provide the {Project} with the data which it needs for its smooth functioning.
+This makes PostgreSQL a heavily used process which if tuned can have a number of benefits on the overall operational response of {Project}.
+
+The below set of tunings can be applied to PostgreSQL to improve its response times (see `how to use custom-hiera.yaml` file; this will modify `/var/lib/pgsql/data/postgresql.conf` file):
```
Applied.

maximiliankolb commented on guides/common/modules/con_postgresql-tuning.adoc:
```diff
@@ -0,0 +1,42 @@
+The cost limit defines the number of tuples that can be processed in a single run by the process.
+An optimal value for this is 2000 based on the general load that {Project} pushes on the PostgreSQL server process.
+
+[NOTE]
+====
+With the upgrade to Postgres 12, `checkpoint_segments` configuration is not supported.
+====
```
Removed.

maximiliankolb commented on guides/common/modules/con_postgresql-tuning.adoc:
```diff
@@ -0,0 +1,42 @@
+
+`autovacuum_vacuum_cost_limit`: The key defines the cost limit value for the vacuuming operation inside the autovacuum process to clean up the dead tuples inside the database relations.
+The cost limit defines the number of tuples that can be processed in a single run by the process.
+An optimal value for this is 2000 based on the general load that {Project} pushes on the PostgreSQL server process.
```
Added.

maximiliankolb commented on guides/common/modules/con_puma-threads.adoc:
```diff
@@ -0,0 +1,16 @@
+[id="Puma_Threads_{context}"]
+= Puma Threads
+
+Less Puma threads lead to more memory usage for different scales on your {ProjectServer}.
```
Applied.

maximiliankolb commented on guides/common/modules/con_smart-proxy-configuration-tuning.adoc:
```diff
@@ -0,0 +1,5 @@
+[id="Smart_Proxy_Configuration_Tuning_{context}"]
+= {SmartProxy} Configuration Tuning
+
+{SmartProxies} are meant to offload part of {Project} load related to distributing content to clients but they can also be used to execute Remote Execution jobs.
```
Reworded.

maximiliankolb commented on guides/common/modules/con_top-performance-considerations.adoc:
```diff
@@ -0,0 +1,20 @@
+You can improve the performance and scalability of {ProjectName}:
+
+. Configure httpd
+. Configure puma to increase concurrency
+. Configure candlepin
+. Configure pulp
```
Applied.

maximiliankolb commented on guides/common/modules/con_top-performance-considerations.adoc:
```diff
@@ -0,0 +1,20 @@
+. Configure Dynflow
+. Deploy external {SmartProxies} instead of relying on internal {SmartProxies}
+. Configure katello-agent for scalability
+. Configure hammer to reduce API timeouts
```
Applied.

maximiliankolb commented on guides/common/modules/proc_benchmark-disk-performance.adoc:
```diff
@@ -0,0 +1,36 @@
+
+[NOTE]
+====
+* One may have to temporarily reduce the RAM in order to run the io benchmark, aka if the box has 256 GiB that is a lot of pulp space, so add mem=20G kernel option in grub.
```
Applied.

maximiliankolb commented on guides/common/modules/proc_benchmark-disk-performance.adoc:
```diff
@@ -0,0 +1,36 @@
+[id="Benchmarking_Disk_Performance_{context}"]
+= Benchmarking Disk Performance
+
+We are working to update `{foreman-maintain}` to only warn the users when its internal quick 'fio' benchmark results in numbers below our recommended throughput but will not require a whitelist parameter to continue.
```
Added the link (& I agree).

maximiliankolb commented:
Rebased to master & resolved various issues. Thanks @ekohl for your review. I am still working on addressing all your comments. **Sorry** if you got a Mail each time I've responded in a thread. I forgot that I should be using the "diff" tab and actually conduct a review.

maximiliankolb commented on guides/common/modules/proc_benchmarking-raw-db-performance.adoc:
```diff
@@ -0,0 +1,27 @@
+To get a list of the top table sizes in disk space for both Candlepin and Foreman, check https://github.com/RedHatSatellite/satellite-support/blob/master/postgres-size-report[postgres-size-report] script in https://github.com/RedHatSatellite/satellite-support[satellite-support] git repository.
+endif::[]
+
+PGbench utility (note you may need to resize PostgreSQL data directory `/var/lib/pgsql/` directory to 100GiB or what does benchmark take to run) might be used to measure PostgreSQL performance on your system.
```
Changed the directory.

maximiliankolb commented on guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qdrouterd.adoc:
```diff
@@ -0,0 +1,23 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+qpid::router::open_file_limit: 150100
```
Described it as default value.

maximiliankolb commented on guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qdrouterd.adoc:
```diff
@@ -0,0 +1,23 @@
+----
+# {foreman-installer}
+# systemctl daemon-reload
+# {foreman-maintain} service restart
```
Applied.

maximiliankolb commented on guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qpidd.adoc:
```diff
@@ -0,0 +1,21 @@
++
+[options="nowrap", subs="+quotes,verbatim,attributes"]
+----
+qpid::open_file_limit: 65536
```
Fixed.

maximiliankolb commented on guides/common/modules/proc_calculating-the-maximum-open-files-limit-for-qpidd.adoc:
```diff
@@ -0,0 +1,21 @@
+----
+# {foreman-installer}
+# systemctl daemon-reload
+# {foreman-maintain} service restart
```
Applied.

maximiliankolb commented on guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc:
```diff
@@ -0,0 +1,28 @@
+apache::mod::prefork::startservers: 10
+----
+
+* Set the *ServerLimit* parameter to raise MaxClients value.
```
Added.

maximiliankolb commented on guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc:
```diff
@@ -0,0 +1,28 @@
+* Set the *ServerLimit* parameter to raise MaxClients value.
+* Set the *MaxClients* parameter to limit the the maximum number of child processes that httpd can launch to handle the incoming requests.
++
+For more information, see `MaxRequestWorker` which is a new name in Apache documentation.
```
Yes, fixed.

maximiliankolb commented on guides/common/modules/proc_configuring-how-many-processes-can-be-launched-by-apache-httpd.adoc:
```diff
@@ -0,0 +1,28 @@
+* Set the *MaxClients* parameter to limit the the maximum number of child processes that httpd can launch to handle the incoming requests.
++
+For more information, see `MaxRequestWorker` which is a new name in Apache documentation.
+* Set the *StartServers* parameter to define how many server processes are launched by default when the httpd process is started.
```
Removed.

maximiliankolb commented on guides/common/modules/proc_configuring-project-environment-for-performance.adoc:
```diff
@@ -0,0 +1,26 @@
+Since {Project} uses PostgreSQL as the database solutions, any additional memory coupled with the tunings will provide a boost to the response times of the applications due to increased data retention in the memory.
+
+Disk::
+With {Project} doing heavy IOPS due to repository synchronizations, package data retrieval, high frequency database updates for the subscription records of the content hosts, it is advised that {Project} be installed on a high speed SSD drive so as to avoid performance bottlenecks which may happen due to increased Disk reads or writes.
```
Applied.

maximiliankolb commented on guides/common/modules/proc_configuring-project-environment-for-performance.adoc:
```diff
@@ -0,0 +1,26 @@
+
+Disk::
+With {Project} doing heavy IOPS due to repository synchronizations, package data retrieval, high frequency database updates for the subscription records of the content hosts, it is advised that {Project} be installed on a high speed SSD drive so as to avoid performance bottlenecks which may happen due to increased Disk reads or writes.
+{Project} requires disk IO to be at or above 60-80 megabytes per second of average throughput for read operations.
```
Added.

maximiliankolb commented on guides/common/modules/proc_configuring-the-maximum-asynchronous-input-output-requests.adoc:
```diff
@@ -0,0 +1,22 @@
+You can increase the maximum number of allowed concurrent AIO requests by increasing the kernel parameter `fs.aio-max-nr`.
+
+.Procedure
+. Set the value of `fs.aio-max-nr` to the desired maximum in `/etc/sysctl.conf`:
```
Applied.

maximiliankolb reviewed (commented):
Thanks so much for all your input! I've resolved some more issues. I will now try to get it to acutally build.

maximiliankolb commented:
This PR is now officially ready for review. :) It aims be to a copy of the content from https://github.com/RedHatSatellite/satellite-performance-tuning/ with minor improvements and corrections.

jhutar commented on guides/common/modules/con_puma-workers-and-threads-recommendation.adoc:
```diff
@@ -0,0 +1,16 @@
+
+[width="100%",cols="16%,19%,7%,7%,31%,20%",options="header",]
+|===
+|Name |Number of managed host |RAM |Cores |Recommended Puma Threads for both min & max |Recommended Puma Workers
+|default |0-5000 |20GiB |4 |16 |4-6
+|medium |5000-10000 |32GiB |8 |16 |8-12
+|large |10000-20000 |64GiB |16|16 |12-18
+|extra-large |20000-60000 |128GiB |32 |16 |16-24
+|extra-extra-large |60000+ |256GiB+ |48+ |16 |20-26
```
Hello Ewoud. That is what we measured - IIRC there was some slight improvement with more threads - we have "Our recommendation is based purely on concurrent registration performance, so it might not reflect your exact use-case." there.

jhutar commented on guides/common/modules/con_postgresql-tuning.adoc:
```diff
@@ -0,0 +1,42 @@
+In the above tuning configuration, there are a certain set of keys which we have altered:
+
+`max_connections`: The key defines the maximum number of connections that can be accepted by the PostgreSQL processes that are running.
+An optimal value for the parameter will be equal to the nearest multiple of 100 of the ServerLimit value of Apache httpd2 multiplied by 2.
+For example, if ServerLimit is set to 582, we can set the max_connections to 1000.
```
Yep, that is old.

maximiliankolb replied:
I need help on this one. cc @jhutar

jhutar replied:
Right, this is very outdated now. What about trimming it to this for now:
```
`max_connections`: The key defines the maximum number of connections that can be accepted by the PostgreSQL processes that are running.
```
I have created https://issues.redhat.com/browse/SAT-10307 to track figuring out more details.

jhutar commented:
Hello. I agree with Ewoud's comments. Could we maybe merge as it is to "snapshot" initial version of the tuning guide and then work on implementing these proposed changes?

maximiliankolb commented:
> Hello. I agree with Ewoud's comments. Could we maybe merge as it is to "snapshot" initial version of the tuning guide and then work on implementing these proposed changes?

That would be awesome! I don't really have the knowledge and time to get into every comment Ewoud made; that why some threads are still open. Of course I can assist with and review future PRs for this guide.

maximiliankolb commented on guides/common/modules/con_postgresql-tuning.adoc:
```diff
@@ -0,0 +1,42 @@
+  max_connections: 1000
+  shared_buffers: 2GiB
+  work_mem: 8MB
+  autovacuum_vacuum_cost_limit: 2000
```
I have claried this option.

maximiliankolb commented on guides/common/modules/con_puma-threads.adoc:
```diff
@@ -0,0 +1,16 @@
+[width="100%",cols="50%,50%",options="header",]
+|===
+|{Project} VM with 8 CPUs, 40 GiB RAM |{Project} VM with 8 CPUs, 40 GiB RAM
+|`--foreman-foreman-service-puma-threads-min=0` |`--foreman-foreman-service-puma-threads-min=16`
+|`--foreman-foreman-service-puma-threads-max=16` |`--foreman-foreman-service-puma-threads-max=16`
+|`--foreman-foreman-service-puma-workers=2` |`--foreman-foreman-service-puma-workers=2`
+|===
+
+Setting the minimum Puma threads to `16` results in about 12% less memory as compared to `t_min=0`.
```
Can you confirm @jhutar? Note that we are providing this only for Foreman 3.1+.

jhutar replied:
Yep, we also recommend min threads == max threads.

We recommended 16 because we had a better results with it.

I agree tuning Puma workers is more straight forward.

This is what I released yesterday as an update to 6.10 tuning guide: https://github.com/RedHatSatellite/satellite-performance-tuning/commit/b87caa34b409aa36c440d9ef044c9fd8ce964ea0#diff-4d49ae92acb7629a016fdcecfa8cfa3458fdda0c93fd14cc1ebebe0cbfe703eeL298-R344

FYI Maximilian, please do not feel obligated to backport all these changes in 6.10 branch. I think we can do these once this huge PR is merged together.

maximiliankolb replied:
Thanks; I've added this!
I will backport this for Foreman 3.1, but obviously Satellite doesn't have to use it right away.

maximiliankolb reviewed (commented):
I have rebased my branch to master again & fixed a couple more issues. I need some more help for the remaining open threads.

maximiliankolb reviewed (commented):
Rebased to master and applied all feedback.

maximiliankolb commented:
Still waiting for review.

jhutar reviewed (approved):
Not sure if my review helps to something, but from mine POV this looks great!

maximiliankolb commented:
Rebased to master once more. I saw no new comments, so I would like to move this ahead and merge this week. Note that we can always fix issues and make things pretty later on. Any objections? cc @Lennonka

Lennonka commented:
Cherry-picked:

- c9943080..fefe65ce  3.3 -> 3.3
- 9f76ea42..b1c9be52  3.2 -> 3.2
- 301c2463..4f3361da  3.1 -> 3.1
